src/toolchain/names.rs

```diff
@@ -248,7 +248,7 @@ impl Display for MaybeOfficialToolchainName {
 /// ToolchainName can be used in calls to Cfg that alter configuration,
 /// like setting overrides, or that depend on configuration, like calculating
 /// the toolchain directory.
-#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
```

Comment:
While indeed the derived `Ord` instance are not used, I guess I was doing something idiomatic here (not entirely my fault, the function has been in existence way before #3518):

https://github.com/rust-lang/rustup/blob/f88394cbbbcc25ffff8e0a0fd465ae77ef22ccbf/src/toolchain/names.rs#L283-L306

How about making use of that to write an actual `PartialOrd` instance?

Reply:
I looked at that but decided to postpone it because I think a `sort()` call that calls a function like this internally will be more expensive? The `.to_string()` call makes this a relatively expensive sorting key, and a `PartialOrd` or `Ord` implementation willl call it twice for each comparison made, whereas `sort_by_key()` can slightly optimize that (although we probably should be using `sort_by_cached_key()` here -- maybe using that directly with moving the key function into a `ToolchainName::sort_key()` method would be a good trade-off for now?).

So I think it would be okay to punt on that for this PR.

Reply:
@djc I agree, it's better if this function is not used in an `Ord` implementation, and something like `Ordering::then_with` should be used instead. Please feel free to move on without this point while I have a closer look into what should be done instead...

Reply:
Split into #3884.